tracing-subscriber/src/registry/fmt.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+// this needs to be a seperate impl block because we're re-assigning the the W2 (make_writer)
+// type paramater from the default.
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W> {
+    /// Sets a [MakeWriter] for spans and events.
+    pub fn with_writer<W2>(self, make_writer: W2) -> FmtLayerBuilder<S, N, E, W2>
+    where
+        W2: MakeWriter + 'static,
+    {
+        FmtLayerBuilder {
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            is_interested: self.is_interested,
+            make_writer,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S, N, E, W> FmtLayerBuilder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    /// Builds a [FmtLayer] infalliably.
+    pub fn build(self) -> FmtLayer<S, N, E, W> {
+        FmtLayer {
+            is_interested: self.is_interested,
+            make_writer: self.make_writer,
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl Default for FmtLayerBuilder {
+    fn default() -> Self {
+        Self {
+            is_interested: Box::new(|_| true),
+            fmt_fields: format::DefaultFields::default(),
+            fmt_event: format::Format::default(),
+            make_writer: io::stdout,
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S, N, E, W> FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    #[inline]
+    fn make_ctx<'a>(&'a self, ctx: Context<'a, S>) -> FmtContext<'a, S, N> {
+        FmtContext {
+            ctx,
+            fmt_fields: &self.fmt_fields,
+        }
+    }
+}
+
+// === impl Formatter ===
+
+impl<S, N, E, W> Layer<S> for FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let span = span.data;
+        let mut fields = String::new();
+        if self.fmt_fields.format_fields(&mut fields, attrs).is_ok() {
+            span.extensions_mut().insert(fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let mut buf = String::new();
+            if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
```

Comment:
This seems wrong to me. `on_record` is recording new fields _on a span that already exists_. I think this should get the span, get the span's extensions, and write the new field to that span's formatted fields extension?

Reply:
Fixed.